### PR TITLE
#692 implementation for EmCatch.Action

### PR DESCRIPTION
**Pull request by dmzaytsev**
For #692 

* created an implementation and unit test for EmCatch.Action which should look for the alias by urn and boat_id and post the message into the bout as the author
* added a puzzle to add header Reply-To to an outgoing email

### CODE REVIEW

**Comment by karato** [comment]
@dmzaytsev Thanks, I will find someone to review this PR soon

**Comment by karato** [comment]
@darkled do this review

File: netbout-web/src/main/java/com/netbout/email/EmAction.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import javax.mail.Message;
+import javax.mail.MessagingException;
+import javax.mail.Multipart;
+
+/**
+ * Post the message into the bout as the author.
+ *
+ * @author Dmitry Zaytsev (dmitry.zaytsev@gmail.com)
+ * @version $Id$
+ * @since 2.16.3
```
Comment (darkled) [inline netbout-web/src/main/java/com/netbout/email/EmAction.java]: @dmzaytsev next version seems to be 2.17

File: netbout-web/src/test/java/com/netbout/email/EmActionTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import com.netbout.spi.Bout;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link EmAction}.
+ * @author Dmitry Zaytsev (dmitry.zaytsev@gmail.com)
+ * @version $Id$
+ * @since 2.16.3
```
Comment (darkled) [inline netbout-web/src/test/java/com/netbout/email/EmActionTest.java]: @dmzaytsev next version seems to be 2.17

File: netbout-web/src/main/java/com/netbout/email/EmAction.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                    final String[] split = EmCatch.decrypt(
+                        adr.substring(0, adr.indexOf('@'))
+                    ).split("\\|");
+                    this.base.user(URN.create(split[0])).aliases()
+                        .iterate().iterator().next().inbox()
+                        .bout(Long.valueOf(split[1]))
+                        .messages().post(this.text(msg));
+                }
+            }
+        } catch (final MessagingException ex) {
```
Comment (darkled) [inline netbout-web/src/main/java/com/netbout/email/EmAction.java]: @dmzaytsev what about surrounding only the lines which could throw `MessagingException` with try-catch?

File: netbout-web/src/main/java/com/netbout/email/EmAction.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+     * @throws IOException If some problem inside
+     */
+    private static String raw(final Message msg) throws IOException {
+        try {
+            final Object body = msg.getContent();
+            if (!(body instanceof Multipart)) {
+                throw new IOException("body is not Multipart");
+            }
+            final Multipart parts = (Multipart) body;
+            for (int pos = 0; pos < parts.getCount(); pos += 1) {
```
Comment (darkled) [inline netbout-web/src/main/java/com/netbout/email/EmAction.java]: @dmzaytsev let's use `++pos`

**Comment by darkled** [comment]
@dmzaytsev please see few minor comments above

File: netbout-web/src/main/java/com/netbout/email/EmAction.java (hunk 3 above)
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/email/EmAction.java]: @darkled  only line 70 
`  : msg.getRecipients(Message.RecipientType.TO)) {` could throw `MessagingException` I have surrounded `for` statement

**Comment by dmzaytsev** [comment]
@darkled done, please take a look

File: netbout-web/src/main/java/com/netbout/email/EmAction.java (hunk 3 above)
Comment (darkled) [inline netbout-web/src/main/java/com/netbout/email/EmAction.java]: @dmzaytsev ok

**Comment by darkled** [comment]
@dmzaytsev thanks

**Comment by darkled** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@rultor deploy

**Comment by karato** [comment]
@ypshenychka please, review this ticket for compliance with our [QA rules](http://www.teamed.io/qa.html)

**Comment by karato** [comment]
@rultor deploy now pls

**Comment by ypshenychka** [comment]
@darkled 
According to our [QA Rules](http://www.teamed.io/qa.html):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

There were two design issues found during code review, and two cosmetic concerns. 
Please confirm that you will try to find at least three major problems next time.

**Comment by darkled** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@darkled Thanks

**Comment by ypshenychka** [comment]
@karato Quality is acceptable.

**Comment by karato** [comment]
> @karato Quality is acceptable.

@ypshenychka thanks for the QA review, we'll work better next time

**Comment by karato** [comment]
@darkled I added 10 mins to @ypshenychka (for QA review) in transaction 63825399; Thanks for your contribution, **19 mins** was added to your account, payment ID is `63825410`, 51 hours and 54 mins spent total; you're getting extra minutes here (c=4); +19 to your rating, your total score is [+234](http://www.netbout.com/b/35091?open=rating)

**Pull request merged**